compile.c

```diff
@@ -4386,9 +4395,27 @@ iseq_optimize(rb_iseq_t *iseq, LINK_ANCHOR *const anchor)
 
             if (do_block_optimization) {
                 INSN * item = (INSN *)list;
-                if (IS_INSN_ID(item, jump)) {
+                // Give up if there is a throw
+                if (IS_INSN_ID(item, throw)) {
                     do_block_optimization = 0;
                 }
+                else {
+                    // If the instruction has a jump target, check if the
+                    // jump target is the block loop label
+                    const char *types = insn_op_types(item->insn_id);
+                    for (int j = 0; types[j]; j++) {
+                        if (types[j] == TS_OFFSET) {
+                            // If the jump target is equal to the block loop
+                            // label, then we can't do the optimization because
+                            // the leading `nop` instruction fires the block
+                            // entry tracepoint
+                            LABEL * target = (LABEL *)OPERAND_AT(item, j);
```

Comment:
Question since this is a bit confusing for me: TS_OFFSET during compilation isn't actually an offset yet, and is an instruction index?

Reply:
Before assembly, it's a pointer to a label.  I'm not sure the best place to link to in order to explain, but if you check [the encoding routine](https://github.com/tenderlove/ruby/blob/aebf96f371c8d874398e0041b798892e545fa206/compile.c#L2651-L2657) when we encounter a TS_OFFSET, it pulls the label instance from the operand and encodes it.  I wanted this patch to be resilient in the case we have new instructions that take a jump target, so rather than checking for specific instructions like `jump`, I check the operands for `TS_OFFSET` and compare that to the block loop label.